pkg/controller/installation/windows/calico_windows_upgrader.go

```diff
@@ -0,0 +1,414 @@
+	return pending, inProgress, inSync, nil
+}
+
+func (w *calicoWindowsUpgrader) upgradeCompleted(node *corev1.Node) bool {
+	_, variant, version := common.GetNodeVariantAndVersion(node)
+	if variant == w.install.Variant && version == w.getExpectedVersion() {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) isUpgradeFromCalicoToEnterprise(node *corev1.Node) bool {
+	_, variant, _ := common.GetNodeVariantAndVersion(node)
+	if variant == operatorv1.Calico && w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) getExpectedVersion() string {
+	if w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return components.EnterpriseRelease
+	}
+	return components.CalicoRelease
+}
+
+func (w *calicoWindowsUpgrader) updateWindowsNodes() {
+	pending, inProgress, inSync, err := w.getNodeUpgradeStatus()
+	if err != nil {
+		windowsLog.Error(err, "Failed to get Windows nodes upgrade status")
+		return
+	}
+
+	// Notify status manager of pending upgrades.
+	for _, node := range pending {
+		w.statusManager.AddWindowsNodeUpgrade(node.Name, false)
+	}
+
+	for _, node := range inProgress {
+		if w.upgradeCompleted(node) {
+			if err := w.finishUpgrade(context.Background(), node); err != nil {
+				// Log the error and continue. We will retry when we update nodes again.
+				windowsLog.Info(fmt.Sprintf("Could not complete upgrade on node %v: %v", node.Name, err))
+				continue
+			}
+			// Successfully finished completing the upgrade. Moving the node
+			// from in-progress to in-sync.
+			inSync[node.Name] = node
+			delete(inProgress, node.Name)
+			w.statusManager.RemoveWindowsNodeUpgrade(node.Name)
+		}
+	}
+
+	// Get the total # of windows nodes we can have upgrading using the
+	// maxUnavailable value, if the node upgrade strategy was respected.
+	numWindowsNodes := len(pending) + len(inProgress) + len(inSync)
+	maxUnavailable, err := intstr.GetValueFromIntOrPercent(w.install.NodeUpdateStrategy.RollingUpdate.MaxUnavailable, numWindowsNodes, false)
+	if err != nil {
+		windowsLog.Error(err, "Invalid maxUnavailable value, falling back to default of 1")
+		maxUnavailable = defaultMaxUnavailable
+	}
+
+	for _, node := range pending {
+		// For upgrades from Calico -> Enterprise, we always upgrade regardless
+		// of maxUnavailable. For other upgrades, check that we have room
+		// available.
+		if w.isUpgradeFromCalicoToEnterprise(node) || len(inProgress) < maxUnavailable {
+			if err := w.startUpgrade(context.Background(), node); err != nil {
+				// Log the error and continue. We will retry when we update nodes again.
+				windowsLog.Info(fmt.Sprintf("Could not start upgrade on node %v: %v", node.Name, err))
+				continue
+			}
+			// Successfully started the upgrade. Moving the node
+			// from pending to in-progress.
+			inProgress[node.Name] = node
+			delete(pending, node.Name)
+			w.statusManager.AddWindowsNodeUpgrade(node.Name, true)
```

Comment:
I don't understand why we need this call - the only difference is that `inProgress` is now set to `true`, however, we've already started the upgrade from within `startUpgrade` called just a few lines above, and all we've done since then is change the contents of a local map. Can't we just get rid of this call?